File: backend/app/api/calendar.py
from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import HTMLResponse

from app.api.deps import verify_token
from app.core.config import GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET
from app.db import calendar_repo, jobs_repo
from app.schemas.calendar import (
    CalendarAccount,
    CalendarAccountCreate,
    CalendarAccountsResponse,
    CalendarProvider,
    CalendarProviderInfo,
    OAuthCallbackRequest,
    OAuthCompleteResponse,
    OAuthStartResponse,
    ProvidersResponse,
)
from app.services.google_oauth import (
    exchange_code_for_tokens,
    get_pending_state,
    get_pending_state_without_remove,
    get_user_email,
    start_oauth_flow,
    SCOPES,
)
from app.services.jobs import enqueue_job

router = APIRouter(prefix="/calendar")


SUPPORTED_PROVIDERS: list[CalendarProviderInfo] = [
    CalendarProviderInfo(
        id=CalendarProvider.google,
        label="Google Calendar",
        auth_method="oauth2_pkce",
        scopes=["https://www.googleapis.com/auth/calendar.readonly"],
        notes="Use installed-app OAuth with loopback redirect and refresh tokens stored locally.",
    ),
    CalendarProviderInfo(
        id=CalendarProvider.apple,
        label="Apple Calendar (CalDAV/iCloud)",
        auth_method="caldav_basic",
        scopes=[],
        notes="Use CalDAV endpoint with app-specific passwords; relies on sync tokens.",
    ),
]


@router.get("/providers", response_model=ProvidersResponse)
async def list_providers(_: None = Depends(verify_token)) -> ProvidersResponse:
    return ProvidersResponse(providers=SUPPORTED_PROVIDERS)


@router.post("/accounts", response_model=CalendarAccount)
async def create_account(
    request: CalendarAccountCreate, _: None = Depends(verify_token)
) -> CalendarAccount:
    account = await calendar_repo.create_account(
        request.provider.value,
        request.display_name,
        request.email,
        request.credentials,
        request.config,
    )
    public_account = calendar_repo.to_public_account(account)
    return CalendarAccount(**public_account)


@router.get("/accounts", response_model=CalendarAccountsResponse)
async def list_accounts(_: None = Depends(verify_token)) -> CalendarAccountsResponse:
    accounts = await calendar_repo.list_accounts(public=True)
    return CalendarAccountsResponse(accounts=[CalendarAccount(**account) for account in accounts])


@router.get("/accounts/{account_id}", response_model=CalendarAccount)
async def get_account(
    account_id: str, _: None = Depends(verify_token)
) -> CalendarAccount:
    account = await calendar_repo.fetch_account(account_id, public=True)
    if not account:
        raise HTTPException(status_code=404, detail="calendar account not found")
    return CalendarAccount(**account)


@router.delete("/accounts/{account_id}")
async def delete_account(account_id: str, _: None = Depends(verify_token)) -> dict:
    account = await calendar_repo.fetch_account(account_id)
    if not account:
        raise HTTPException(status_code=404, detail="calendar account not found")
    await calendar_repo.delete_account(account_id)
    return {"ok": True}


@router.post("/accounts/{account_id}/sync")
async def sync_account(
    account_id: str,
    payload: dict,
    _: None = Depends(verify_token),
) -> dict:
    account = await calendar_repo.fetch_account(account_id)
    if not account:
        raise HTTPException(status_code=404, detail="calendar account not found")
    job_id = await jobs_repo.create_job(
        "connector.calendar.sync", {"account_id": account_id, **payload}
    )
    await enqueue_job(job_id)
    return {"job_id": job_id, "status": "queued"}


# =============================================================================
# Google OAuth Endpoints
# =============================================================================


@router.post("/oauth/google/start", response_model=OAuthStartResponse)
async def start_google_oauth(
    request: Request,
    _: None = Depends(verify_token),
) -> OAuthStartResponse:
    """
    Start Google OAuth flow.
    Returns auth URL that frontend should open in user's default browser.
    """
    if not GOOGLE_CLIENT_ID or not GOOGLE_CLIENT_SECRET:
        raise HTTPException(
            status_code=500,
            detail="Google OAuth not configured. Set GOOGLE_CLIENT_ID and GOOGLE_CLIENT_SECRET.",
        )

    # Get the backend port from the request
    backend_port = request.url.port or 80

    try:
        result = start_oauth_flow(backend_port)
        return OAuthStartResponse(**result)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/oauth/google/callback", response_class=HTMLResponse)
async def google_oauth_callback(
    code: str = None,
    state: str = None,
    error: str = None,
):
    """
    Handle OAuth callback from Google.
    This is called by Google after user authorizes.
    """
    # Error from Google
    if error:
        return HTMLResponse(
            content=f"""
            <!DOCTYPE html>
            <html>
            <head><title>Authorization Failed</title></head>
            <body style="font-family: system-ui; display: flex; justify-content: center; align-items: center; height: 100vh; margin: 0;">
                <div style="text-align: center;">
                    <h1 style="color: #dc2626;">Authorization Failed</h1>
                    <p>Error: {error}</p>
                    <p>You can close this window.</p>
                </div>
            </body>
            </html>
            """,
            status_code=400,
        )

    if not code or not state:
        return HTMLResponse(
            content="""
            <!DOCTYPE html>
            <html>
            <head><title>Authorization Failed</title></head>
            <body style="font-family: system-ui; display: flex; justify-content: center; align-items: center; height: 100vh; margin: 0;">
                <div style="text-align: center;">
                    <h1 style="color: #dc2626;">Authorization Failed</h1>
                    <p>Missing authorization code or state.</p>
                    <p>You can close this window.</p>
                </div>
            </body>
            </html>
            """,
            status_code=400,
        )

    # Retrieve pending state
    state_data = get_pending_state(state)
    if not state_data:
        return HTMLResponse(
            content="""
            <!DOCTYPE html>
            <html>
            <head><title>Authorization Failed</title></head>
            <body style="font-family: system-ui; display: flex; justify-content: center; align-items: center; height: 100vh; margin: 0;">
                <div style="text-align: center;">
                    <h1 style="color: #dc2626;">Authorization Failed</h1>
                    <p>Invalid or expired OAuth state. Please try again.</p>
                    <p>You can close this window.</p>
                </div>
            </body>
            </html>
            """,
            status_code=400,
        )

    try:
        # Exchange authorization code for tokens
        tokens = await exchange_code_for_tokens(
            code=code,
            code_verifier=state_data["code_verifier"],
            redirect_uri=state_data["redirect_uri"],
        )

        # Get user email from Google
        email = await get_user_email(tokens["access_token"])

        # Store credentials
        credentials_payload = {
            "access_token": tokens["access_token"],
            "refresh_token": tokens.get("refresh_token"),
            "expires_in": tokens.get("expires_in"),
        }

        # Create calendar account
        await calendar_repo.create_account(
            provider="google",
            display_name=email or "Google Calendar",
            email=email,
            credentials=credentials_payload,
            config={"scopes": SCOPES},
        )

        return HTMLResponse(
            content=f"""
            <!DOCTYPE html>
            <html>
            <head><title>Authorization Successful</title></head>
            <body style="font-family: system-ui; display: flex; justify-content: center; align-items: center; height: 100vh; margin: 0; background: #f0fdf4;">
                <div style="text-align: center;">
                    <h1 style="color: #16a34a;">Authorization Successful!</h1>
                    <p>Google Calendar ({email or 'account'}) has been connected.</p>
                    <p style="color: #666;">You can close this window and return to the app.</p>
                </div>
            </body>
            </html>
            """
        )

    except Exception as e:
        return HTMLResponse(
            content=f"""
            <!DOCTYPE html>
            <html>
            <head><title>Authorization Failed</title></head>
            <body style="font-family: system-ui; display: flex; justify-content: center; align-items: center; height: 100vh; margin: 0;">
                <div style="text-align: center;">
                    <h1 style="color: #dc2626;">Authorization Failed</h1>
                    <p>Error: {str(e)}</p>
                    <p>You can close this window and try again.</p>
                </div>
            </body>
            </html>
            """,
            status_code=500,
        )


@router.post("/oauth/google/complete", response_model=OAuthCompleteResponse)
async def complete_google_oauth(
    request: OAuthCallbackRequest,
    _: None = Depends(verify_token),
) -> OAuthCompleteResponse:
    """
    Complete OAuth flow by exchanging code for tokens.
    Called by frontend after receiving callback from Google.
    (This is now optional - the callback endpoint handles everything)
    """
    # Retrieve and validate pending state
    state_data = get_pending_state(request.state)
    if not state_data:
        raise HTTPException(status_code=400, detail="Invalid or expired OAuth state")

    try:
        # Exchange authorization code for tokens
        tokens = await exchange_code_for_tokens(
            code=request.code,
            code_verifier=state_data["code_verifier"],
            redirect_uri=state_data["redirect_uri"],
        )
    except Exception as e:
        raise HTTPException(status_code=400, detail=f"Token exchange failed: {str(e)}")

    # Get user email from Google
    email = await get_user_email(tokens["access_token"])

    # Store credentials
    credentials_payload = {
        "access_token": tokens["access_token"],
        "refresh_token": tokens.get("refresh_token"),
        "expires_in": tokens.get("expires_in"),
    }

    # Create calendar account
    account = await calendar_repo.create_account(
        provider="google",
        display_name=email or "Google Calendar",
        email=email,
        credentials=credentials_payload,
        config={"scopes": SCOPES},
    )

    public_account = calendar_repo.to_public_account(account)

    return OAuthCompleteResponse(
        account=CalendarAccount(**public_account),
        success=True,
        message="Google Calendar connected successfully",
    )

File: backend-bun/src/routes/calendar.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";
import { authMiddleware, requireAuth } from "../middleware/auth.ts";
import * as calendarRepo from "../db/calendar.ts";
import * as jobsRepo from "../db/jobs.ts";
import { enqueueJob } from "../services/jobs.ts";
import {
  startGoogleOAuth,
  getOAuthState,
  exchangeCodeForTokens,
  fetchUserEmail,
} from "../services/calendar/oauth.ts";
import { manager } from "../websocket/manager.ts";
import { parseIsoToEpoch } from "../lib/time.ts";
import {
  CalendarAccountCreateSchema,
  CalendarSelectionUpdateSchema,
  OAuthCallbackRequestSchema,
  type CalendarProviderInfo,
  type OAuthStartResponse,
  type OAuthCompleteResponse,
} from "../schemas/calendar.ts";
import type { Env } from "../lib/types.ts";

const PROVIDERS: CalendarProviderInfo[] = [
  {
    id: "google",
    label: "Google Calendar",
    auth_method: "oauth2_pkce",
    scopes: ["https://www.googleapis.com/auth/calendar.readonly"],
    notes: "Uses OAuth 2.0 with PKCE for secure desktop app authentication",
  },
  {
    id: "apple",
    label: "Apple Calendar",
    auth_method: "caldav",
    scopes: [],
    notes: "Uses CalDAV protocol with app-specific password",
  },
];

export const calendarRoutes = new Hono<Env>();

calendarRoutes.use("*", authMiddleware);

// Providers (requires auth)
calendarRoutes.get("/providers", requireAuth, (c) => {
  return c.json({ providers: PROVIDERS });
});

// Accounts
calendarRoutes.post(
  "/accounts",
  requireAuth,
  zValidator("json", CalendarAccountCreateSchema),
  (c) => {
    const body = c.req.valid("json");
    const accountId = calendarRepo.createAccount(
      body.provider,
      body.display_name ?? null,
      body.email ?? null,
      body.credentials ?? null,
      body.config ?? null
    );
    const account = calendarRepo.fetchAccount(accountId);
    if (!account) throw new Error("Failed to create account");
    const { credentials: _, ...publicAccount } = account;
    return c.json(publicAccount);
  }
);

calendarRoutes.get("/accounts", requireAuth, (c) => {
  return c.json({ accounts: calendarRepo.fetchAccounts() });
});

calendarRoutes.get(
  "/accounts/:accountId",
  requireAuth,
  zValidator("param", z.object({ accountId: z.string() })),
  (c) => {
    const { accountId } = c.req.valid("param");
    const account = calendarRepo.fetchAccount(accountId);
    if (!account) {
      return c.json({ detail: "Account not found" }, 404);
    }
    const { credentials: _, ...publicAccount } = account;
    return c.json(publicAccount);
  }
);

calendarRoutes.delete(
  "/accounts/:accountId",
  requireAuth,
  zValidator("param", z.object({ accountId: z.string() })),
  (c) => {
    const { accountId } = c.req.valid("param");
    calendarRepo.deleteAccount(accountId);
    return c.json({ ok: true });
  }
);

// Calendars
calendarRoutes.get(
  "/calendars",
  requireAuth,
  zValidator(
    "query",
    z.object({
      account_id: z.string().optional(),
      selected_only: z.string().optional(),
    })
  ),
  (c) => {
    const query = c.req.valid("query");
    const accountId = query.account_id;
    const selectedOnly = query.selected_only !== "false";
    return c.json({ calendars: calendarRepo.fetchCalendars(accountId, selectedOnly) });
  }
);

calendarRoutes.patch(
  "/calendars/:accountId/:calendarId",
  requireAuth,
  zValidator("param", z.object({ accountId: z.string(), calendarId: z.string() })),
  zValidator("json", CalendarSelectionUpdateSchema),
  (c) => {
    const { accountId, calendarId } = c.req.valid("param");
    const body = c.req.valid("json");
    calendarRepo.updateCalendarSelection(accountId, calendarId, body.selected);
    return c.json({ ok: true });
  }
);

// Events
calendarRoutes.get(
  "/events",
  requireAuth,
  zValidator(
    "query",
    z.object({
      start: z.string(),
      end: z.string(),
      account_id: z.string().optional(),
      calendar_ids: z.string().optional(),
      selected_only: z.string().optional(),
    })
  ),
  (c) => {
    const query = c.req.valid("query");

    const startTs = parseIsoToEpoch(query.start);
    const endTs = parseIsoToEpoch(query.end);

    if (startTs >= endTs) {
      return c.json({ detail: "start must be before end" }, 400);
    }

    const calendarIds = query.calendar_ids?.split(",").filter(Boolean);
    const selectedOnly = query.selected_only !== "false";

    return c.json({
      events: calendarRepo.fetchEvents(startTs, endTs, {
        accountId: query.account_id,
        calendarIds,
        selectedOnly,
      }),
    });
  }
);

// Sync
calendarRoutes.post(
  "/accounts/:accountId/sync",
  requireAuth,
  zValidator("param", z.object({ accountId: z.string() })),
  zValidator("json", z.record(z.string(), z.unknown()).optional()),
  (c) => {
    const { accountId } = c.req.valid("param");
    const body = c.req.valid("json") ?? {};
    const payload = { ...body, account_id: accountId };
    const jobId = jobsRepo.createJob("connector.calendar.sync", payload as Record<string, unknown>);
    enqueueJob(jobId);
    manager.emitLog("info", `Calendar sync job queued ${jobId}`);
    return c.json({ job_id: jobId, status: "queued" });
  }
);

// Google OAuth
calendarRoutes.post("/oauth/google/start", requireAuth, async (c) => {
  const result = await startGoogleOAuth();
  return c.json<OAuthStartResponse>(result);
});

calendarRoutes.get(
  "/oauth/google/callback",
  zValidator(
    "query",
    z.object({
      code: z.string().optional(),
      state: z.string().optional(),
      error: z.string().optional(),
    })
  ),
  async (c) => {
    const { code, state, error } = c.req.valid("query");

    if (error) {
      return c.html(
        `<html><body><h1>Error</h1><p>${error}</p><script>window.close();</script></body></html>`
      );
    }

    if (!code || !state) {
      return c.html(
        `<html><body><h1>Error</h1><p>Missing code or state</p><script>window.close();</script></body></html>`
      );
    }

    const oauthState = getOAuthState(state);
    if (!oauthState) {
      return c.html(
        `<html><body><h1>Error</h1><p>Invalid or expired state</p><script>window.close();</script></body></html>`
      );
    }

    try {
      const tokens = await exchangeCodeForTokens(
        code,
        oauthState.code_verifier,
        oauthState.redirect_uri
      );

      const email = await fetchUserEmail(tokens.access_token);

      calendarRepo.createAccount(
        "google",
        email,
        email,
        tokens,
        { scopes: ["https://www.googleapis.com/auth/calendar.readonly"] }
      );

      manager.emitLog("info", `Google Calendar connected: ${email}`);

      return c.html(
        `<html><body><h1>Success!</h1><p>Google Calendar connected as ${email}</p><p>You can close this window.</p><script>window.close();</script></body></html>`
      );
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : "Unknown error";
      return c.html(
        `<html><body><h1>Error</h1><p>${errorMessage}</p><script>window.close();</script></body></html>`
      );
    }
  }
);

calendarRoutes.post(
  "/oauth/google/complete",
  requireAuth,
  zValidator("json", OAuthCallbackRequestSchema),
  async (c) => {
    const body = c.req.valid("json");
    const oauthState = getOAuthState(body.state);
    if (!oauthState) {
      return c.json<OAuthCompleteResponse>({
        account: null,
        success: false,
        message: "Invalid or expired state",
      });
    }

    try {
      const tokens = await exchangeCodeForTokens(
        body.code,
        oauthState.code_verifier,
        oauthState.redirect_uri
      );

      const email = await fetchUserEmail(tokens.access_token);

      const accountId = calendarRepo.createAccount(
        "google",
        email,
        email,
        tokens,
        { scopes: ["https://www.googleapis.com/auth/calendar.readonly"] }
      );

      const account = calendarRepo.fetchAccount(accountId);
      if (!account) {
        return c.json<OAuthCompleteResponse>({
          account: null,
          success: false,
          message: "Failed to create account",
        });
      }

      const { credentials: _, ...publicAccount } = account;
      return c.json<OAuthCompleteResponse>({
        account: publicAccount,
        success: true,
        message: "Success",
      });
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : "Unknown error";
      return c.json<OAuthCompleteResponse>({
        account: null,
        success: false,
        message: errorMessage,
      });
    }
  }
);
